doc/rtd/explanation/failure_states.rst

```diff
@@ -0,0 +1,126 @@
+.. _failure_states:
+
+Failure states
+==============
+
+Critical failure
+----------------
+If cloud-init is unable to complete, error messages will now
+additionally be visible in output of `cloud-init status --long`
+within the 'error' key nested under the module-level keys: 'init-local',
+'init', 'modules-config', 'modules-final'.
+
+Recoverable failure
+-------------------
+In the case that cloud-init is able to complete yet something goes awry,
+the service will now return with exit code 2, and error messages will be
+visible in the output of `cloud-init status --long` under the top
+level 'recoverable_errors' key as well as within the 'error' key nested
+under the module-level keys: 'init-local', 'init', 'modules-config',
+'modules-final'.
+
+Implementation
+==============
+
+Cloud-init error codes
+----------------------
+0 - success
+1 - unrecoverable error
+2 - recoverable error
+
+If cloud-init exits with exit code 1, cloud-init experienced critical failure
+and was unable to recover. In this case, something is likely seriously
+wrong with the system, or cloud-init has experienced a serious bug.
+
+If cloud-init exits with exit code 2, cloud-init was able to complete
+gracefully, however something went wrong and the user should investigate.
+
+
+Reported state
+--------------
+Cloud-init can report its internal state via the `status --format json`
+subcommand under the `'extended_status'` key.
+
+.. code-block: shell-session
+
+    $ cloud-init status --format json | jq .extended_status
+    "degraded done"
+
+See the list of all possible states:
+
+.. code-block: shell-session
+
+  "not running"
+  "running"
+  "done"
+  "error"
+  "degraded done"
+  "degraded running"
+  "disabled"
+
+
+Exported errors: Aggregated errors
+----------------------------------
+When a recoverable error occurs, the internal cloud-init state
```

Comment:
> We haven't defined anywhere what a "recoverable error" is 

See the "error codes" section

> and it doesn't match the section heading, either.

This section only applies to recoverable errors, not unrecoverable errors, should that be clarified maybe?

Reply:
Looking back to where you pointed, I think the definition of a recoverable error is fine (although thinking about it, I'd maybe put subsubheadings for the recoverable/non-recoverable error definitions up there to call those sections out specifically). The primary confusion for me is how recoverable/non-recoverable errors relate to Exported and Aggregated errors (as in the headings) since this isn't immediately obvious.

Reply:
The exported and aggregated errors content relates to recoverable errors. I've broken this content out into a separate page to make that more clear.